src/osd/PGLog.h

```diff
@@ -1112,19 +1112,27 @@ struct PGLog : DoutPrefixProvider {
 			   << "," << info.last_update << "]" << dendl;
 
 	set<hobject_t, hobject_t::BitwiseComparator> did;
+	set<hobject_t, hobject_t::BitwiseComparator> del;
 	set<hobject_t, hobject_t::BitwiseComparator> checked;
 	for (list<pg_log_entry_t>::reverse_iterator i = log.log.rbegin();
 	     i != log.log.rend();
 	     ++i) {
-	  if (!debug_verify_stored_missing && i->version <= info.last_complete) break;
 	  if (cmp(i->soid, info.last_backfill, info.last_backfill_bitwise) > 0)
 	    continue;
 	  if (i->is_error())
 	    continue;
+	  // merge clean_regions in pg_log_entry_t if needed
```

Comment:
@athanatos , urh， yeah，you are right, current master modify read_log/write_log to read_log_and_missing/write_log_and_missing.  So it is used to check whether our missing is correct as expected based on the flag debug_verify_stored_missing.

And also, in this way, we need to consider retrieve pg_missing_item from the disk as you mention before. I think we can add a tail on the key such that p->key().substr(0, 7) == string("missing") && p->key().substr(7, 9) == string("v1")  then decode new version and else decode old versions? Do you think this is reasonable?

Also I agree that it is much easier to deal with cases here if encode/decode is wrapped by the MACRO. But as discussed before, is that worthwhile to change the interface of the decode so that we can also pass the feature bit into the decode function?